### ISSUE TITLE
file_taint segmentation fault

### ISSUE THREAD

**Issue by Toizi** [panda-re/panda#607]
I am using the image from http://panda-re.mit.edu/qcows/linux/ubuntu/1804/x86_64/bionic-server-cloudimg-amd64-noaslr-nokaslr.qcow2 and attempt to use the `file_taint` with `tainted_branch` plugin on a simple example of just `cat`ing a file. What I get is the following output (`tainted_branch` is not needed to trigger the segfault)
```
$ build/x86_64-softmmu/panda-system-x86_64 -m 4096 -replay cat_test -panda file_taint:filename=/root/osi_linux/kernelinfo.conf -panda osi -os linux-64-ubuntu:4.15.0-72-generic-noaslr-nokaslr       
PANDA[file_taint]:adding argument filename=/root/osi_linux/kernelinfo.conf.
PANDA[core]:os_familyno=2 bits=64 os_details=ubuntu:4.15.0-72-generic-noaslr-nokaslr
PANDA[core]:initializing file_taint
PANDA[core]:loading required plugin syscalls2
PANDA[core]:initializing syscalls2
PANDA[syscalls2]:using profile for linux x64 64-bit
PANDA[core]:loading required plugin osi
PANDA[core]:initializing osi
PANDA[core]:loading required plugin osi_linux
PANDA[core]:initializing osi_linux
PANDA[osi_linux]:W> kernelinfo bytes [20-23] not read
PANDA[osi_linux]:W> kernelinfo bytes [124-127] not read
PANDA[core]:loading required plugin taint2
PANDA[core]:initializing taint2
PANDA[taint2]:propagation via pointer dereference ENABLED
PANDA[taint2]:taint operations inlining DISABLED
PANDA[taint2]:llvm optimizations DISABLED
PANDA[taint2]:taint debugging DISABLED
PANDA[taint2]:detaint if control bits 0 DISABLED
PANDA[taint2]:maximum taint compute number (0=unlimited) 0
PANDA[taint2]:maximum taintset cardinality (0=unlimited) 0
PANDA[core]:loading required plugin callstack_instr
PANDA[core]:initializing callstack_instr
callstack_instr:  setting up threaded stack_type
PANDA[core]:loading required plugin osi
PANDA[core]:/home/user/dev/panda/build/x86_64-softmmu/panda/plugins/panda_osi.so already loaded
PANDA[core]:loading required plugin osi_linux
PANDA[core]:/home/user/dev/panda/build/x86_64-softmmu/panda/plugins/panda_osi_linux.so already loaded
PANDA[core]:/home/user/dev/panda/build/x86_64-softmmu/panda/plugins/panda_osi.so already loaded
loading snapshot
... done.
opening nondet log for read :	./cat_test-rr-nondet.log
./cat_test-rr-nondet.log:    10838260 instrs total.
cat_test:      108397 (  1.00%) instrs.    0.29 sec.  0.50 GB ram.
cat_test:      216778 (  2.00%) instrs.    0.35 sec.  0.51 GB ram.
cat_test:      325148 (  3.00%) instrs.    0.40 sec.  0.51 GB ram.
cat_test:      433533 (  4.00%) instrs.    0.48 sec.  0.51 GB ram.
cat_test:      541914 (  5.00%) instrs.    0.55 sec.  0.51 GB ram.
cat_test:      650300 (  6.00%) instrs.    0.59 sec.  0.51 GB ram.
cat_test:      758682 (  7.00%) instrs.    0.65 sec.  0.51 GB ram.
cat_test:      867068 (  8.00%) instrs.    0.70 sec.  0.51 GB ram.
cat_test:      975444 (  9.00%) instrs.    0.76 sec.  0.51 GB ram.
cat_test:     1083832 ( 10.00%) instrs.    0.81 sec.  0.51 GB ram.
cat_test:     1192212 ( 11.00%) instrs.    0.85 sec.  0.51 GB ram.
cat_test:     1300593 ( 12.00%) instrs.    0.91 sec.  0.51 GB ram.
cat_test:     1408974 ( 13.00%) instrs.    0.97 sec.  0.51 GB ram.
cat_test:     1517360 ( 14.00%) instrs.    1.03 sec.  0.51 GB ram.
cat_test:     1625741 ( 15.00%) instrs.    1.07 sec.  0.51 GB ram.
cat_test:     1734127 ( 16.00%) instrs.    1.12 sec.  0.51 GB ram.
cat_test:     1842507 ( 17.00%) instrs.    1.18 sec.  0.52 GB ram.
cat_test:     1950889 ( 18.00%) instrs.    1.24 sec.  0.52 GB ram.
cat_test:     2059276 ( 19.00%) instrs.    1.29 sec.  0.52 GB ram.
cat_test:     2167658 ( 20.00%) instrs.    1.34 sec.  0.52 GB ram.
cat_test:     2276037 ( 21.00%) instrs.    1.38 sec.  0.52 GB ram.
file_taint read_enter: first time match of file "/root/osi_linux/kernelinfo.conf", enabling taint
PANDA[taint2]:taint2_enable_taint
taint2: Allocating large fast_shad (68719476736 bytes).
taint2: Allocating small fast_shad (12800000 bytes) using malloc @ 0x7f1b90bca010.
taint2: Allocating small fast_shad (256 bytes) using malloc @ 0x7f1b8cce9370.
taint2: Allocating small fast_shad (2048 bytes) using malloc @ 0x7f1b8cdf0500.
taint2: Allocating small fast_shad (686848 bytes) using malloc @ 0x7f1b93664010.
PANDA[taint2]:LLVM optimizations DISABLED
taint2: Linking taint ops from /home/user/dev/panda/build/x86_64-softmmu/panda/plugins/taint2/panda_taint2_ops.bc
taint2: Done initializing taint transformation.
PANDA[taint2]:Done processing helper functions for taint.
Done verifying module. Running...
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
cat_test:     2384420 ( 22.00%) instrs.    7.34 sec.  0.70 GB ram.
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
warning: encountered a value greater than 64 bits - not attempting to propagate taint through mul instruction
*** applying uniform taint labels 0..12965 to buffer @ 2420755 ***
PANDA[taint2]: Fatal error- taint query on invalid address 0x12c0d1000
[1]    24537 segmentation fault (core dumped)  build/x86_64-softmmu/panda-system-x86_64 -m 4096 -replay cat_test -panda   os
```

To reproduce
(qcow image at `~/dev/qemu_disks/bionic-server-cloudimg-amd64-noaslr-nokaslr.qcow2`):
- boot with
`build/x86_64-softmmu/panda-system-x86_64 -boot c -m 4096 -hda ~/dev/qemu_disks/bionic-server-cloudimg-amd64-noaslr-nokaslr.qcow2 -monitor stdio`
- login with `root:root`
- in monitor `begin_record cat_test`
- in guest `cat /root/osi_linux/kernelinfo.conf`
- in monitor `end_record`, then `quit`
- now use `taint_file` plugin to taint the file we read with `cat` by using
`build/x86_64-softmmu/panda-system-x86_64 -m 4096 -replay cat_test -panda file_taint:filename=/root/osi_linux/kernelinfo.conf -panda osi -os linux-64-ubuntu:4.15.0-72-generic-noaslr-nokaslr`

Edit: Panda version is current master, i.e. `96083243b9d1b5ba0135c2603227f8960b8cbc36`

**Comment by AndrewFasano** [comment]
Hm, I added that "Fatal error- taint query on invalid address" message and logic to bail when it happens fairly recently (in 5644b00e8b1369f9827225054d5032e8543cb8a2) but I think that was correct. It looks like the taint system is somehow querying an invalid address.

I tried the to reproduce with the same setup as you, but I changed the args to use `-m 1024` (because then you can do `loadvm root` from the monitor to skip past boot) and that seemed to work fine (at least it didn't crash).

I'll try again from boot with 4g ram and see if that causes the issue

**Comment by AndrewFasano** [comment]
With `-m 4096` I'm able to reproduce the crash.

**Comment by AndrewFasano** [comment]
We appear to have a bug in our taint system when using >3G ram, it's related to the [3GB barrier](https://en.wikipedia.org/wiki/3_GB_barrier) and we have some ideas of how to fix it. Should have a patch for this soon.

**Comment by Toizi** [comment]
Sounds good. Thanks for taking the time to look into this @AndrewFasano

**Comment by AndrewFasano** [comment]
Should be fixed after #654
